### DISCUSSION TITLE
Updating Marker set

### DISCUSSION THREAD

**Discussion by heinwinter** [iTwin/itwinjs-core#3512] [Q&A]
Hi,
I have incoming real time data that I get from an api. I've been trying to show this data on my iModel while refreshing the data every 10 seconds which I can do, but now my icons refresh with the data so they keep on flashing on my iModel (The whole decorator gets replaced). I want to keep the icons static while refreshing the cloud data in the tooltip
![73ea84df-84c5-4ff0-bf24-54e708003b71](https://user-images.githubusercontent.com/103569255/163147108-aeb62b72-bf85-4f91-8740-0d224d947286.jpg)
I have tried using the code snippet above. As well as various different methods(attached below) with no success.

private async addIcons()
    {
        const values = await SmartDeviceDecorator.getSmartDeviceData();
        values.forEach(value =>{
            this._markerSet.forEach(marker =>{
                marker.setImageUrl(`/${value.smartDeviceType}.png`);
            })
        });
}
private async updateCloudData(){
        const values = await SmartDeviceDecorator.getSmartDeviceData();
        const cloudData = await SmartDeviceAPI.getData();

        this._markerSet.forEach(marker=>{
            setInterval(()=>{
                values.forEach(value =>{
                    marker.title = this.populateTitle(cloudData,value.smartDeviceId);
                }) 
            }, 5000);
            cloudData.push(marker)
        }); 
    }

I know that I can iterate through the marketSet to update the cloudData but I'm unsure on how to complete this action or which object inside the marker set to update. Any help would be appreciated.

**Comment by ajay-toobler** [comment]
have u got solution for this ?

**Comment by pmconne** [comment]
I am certain there was another discussion about a similar issue, in which a user was trying to dynamically update the contents of the marker tooltip while hovering, but I cannot find it.
`Marker.title` can be a string **or** an HTMLElement. In the latter case, if you were to change the contents (e.g., `innerText`) of the HTMLElement while hovering you'd reasonably expect it to update on-screen immediately.
[This sample](https://www.itwinjs.org/sandboxes/pmconne/dynamic-marker-tooltips) attempts to do that - see its README for details.
It does not work. @NancyMcCallB I believe this is an issue with the React-based UI - is there a solution?

**Reply by ajay-toobler** [reply]
That sample too not showing the result as expected. changed title value only display when mouse consistently move after hovered.

I have seen https://www.youtube.com/watch?v=gDFXaOVBepI&t=133s this video working fine. Can you help me to findout how they implemented this.?

**Reply by pmconne** [reply]
Yes that's the whole point of the sample, to illustrate the problem.
Perhaps @roopksaini can tell you how the tooltips in that demo work.

**Reply by NancyMcCallB** [reply]
The issue is that to be passed as an HTMLElement, a React component must render to HTML. At that point is is a snapshot of the sate of the React component, not an actual React component, so making changes to the state will not cause it to re-render. The only way I can think of to fully support a component that's not plain HTML is to change those types (tooltips, titles, etc.) to type any.

**Reply by MichaelBelousov** [reply]
The (~~rather old 2.x~~ apparently it was updated to 3.x) [@bentley/imodel-react-hooks package](https://www.npmjs.com/package/@bentley/imodel-react-hooks) implements markers that re-render on state changes in user react code (see gif below).
Has the marker API changed significantly since then? The code it uses is [here](https://github.com/iTwin/viewer-components-react/blob/651d8e6cd772ced3a6e3b48623f86035bb6bb220/packages/itwin/imodel-react-hooks/src/Marker/useMarker.ts#L209).

There's even a gif of it working on hover in [a blog post](https://medium.com/itwinjs/introducing-bentley-imodel-react-hooks-8945a72711d2).
![image](https://miro.medium.com/max/143/1*xW7lBsjxzLhLLcnazAq74w.gif)

I am not certain if it will work completely in this case

**Reply by tonygem2022** [reply]
Good thread. Similar issues we have.
The other one is how can we make the markers occluded?
Currently we see all markers regardless of the user's view. We would like to have the markers occluded if they are behind an object and only show those markers visible between the user and the object, not behind.
Guidance welcomed or reference to another thread perhaps. 

Thanks. Tony

**Reply by ajay-toobler** [reply]
It could be better If I get respone from @roopksaini about the working of tooltip mentioned in the about youtube link.

**Reply by ajay-toobler** [reply]
I have done a hack.

I have observed that the title updated when mouse move. So I used a private property to store latest mouse move event. so when I needed to update the title I am just calling the `onMouseMove` of parent with the latest mouse move event.

```

export class SmartDeviceMarker extends Marker {
  private _smartDeviceId: string;
  private _smartDeviceType: string;
  private _EcInstanceId: string;
  private _mouseMoveLatestEvent?: BeButtonEvent;

  constructor(
    location: XYAndZ,
    size: XAndY,
    smartDeviceId: string,
    smartDeviceType: string,
    cloudData: any,
    EcInstanceId: any
  ) {
    super(location, size);

    this._smartDeviceId = smartDeviceId;
    this._smartDeviceType = smartDeviceType;
    this._EcInstanceId = EcInstanceId;
    this._mouseMoveLatestEvent = undefined;

    this.setImageUrl(`/${this._smartDeviceType}.png`);
  }

  onMouseMove(ev: BeButtonEvent): void {
    this._mouseMoveLatestEvent = ev;
    super.onMouseMove(ev);
  }

  onMouseLeave(): void {
    this._mouseMoveLatestEvent = undefined;
    super.onMouseLeave();
  }

  updateTitle(cloudData: any) {
    this.title = this.populateTitle(cloudData);
    if (this._mouseMoveLatestEvent && this._isHilited) {
      super.onMouseMove(this._mouseMoveLatestEvent);
    }
  }

  updateColor(cloudData: any) {
    ElementColorStatus.updateDeviceColorStatus(cloudData, {
      id: this._EcInstanceId,
      smartDeviceId: this._smartDeviceId,
    });
    if (this._mouseMoveLatestEvent && this._isHilited) {
      super.onMouseMove(this._mouseMoveLatestEvent);
    }
  }

  private populateTitle(cloudData: any) {
    let smartTable = "";
    for (const [key, value] of Object.entries(cloudData)) {
      smartTable += `
        <tr>
          <th>${key}</th>
          <th>${value}</th>
        </tr>
      `;
    }

    const smartTableDiv = document.createElement("div");
    smartTableDiv.className = "smart-table";
    smartTableDiv.innerHTML = `
     <h3>${this._smartDeviceId}</h3>
     <table>
      ${smartTable}
     </table>
    `;

    return smartTableDiv;
  }
}


```

Now it's working as I expected. Let me know am I good to go with this or not !

**Reply by pmconne** [reply]
Please don't do that. Please do read @MichaelBelousov's [article](https://medium.com/itwinjs/introducing-bentley-imodel-react-hooks-8945a72711d2). I have asked @roopksaini to provide details about the demo recording you found.

@tonygem2022 please start a separate discussion for your question.

**Reply by Josh-Schifter** [reply]
> Good thread. Similar issues we have. The other one is how can we make the markers occluded? Currently we see all markers regardless of the user's view. We would like to have the markers occluded if they are behind an object and only show those markers visible between the user and the object, not behind. Guidance welcomed or reference to another thread perhaps.

You can draw the markers as World Decorations.  This [sandbox](https://www.itwinjs.org/sandboxes/ThomasDicarlo/World%20Decoration%20Markers) shows how to use a combination of world decorations and world overlays so that you can still see a faint indication of the markers even when they are occluded:


Also, see [GraphicType ](https://www.itwinjs.org/reference/core-frontend/rendering/graphictype/) for the various ways to draw decorators.

**Reply by roopksaini** [reply]
@ajay-toobler: Here's the exact [class](https://github.com/iModeljs-meets-AzureDT/civil-iot/blob/master/viewer/src/frontend/components/SensorMarker.tsx) used for  marker tooltips (in the IoT demo).
The repository is outdated (from 2020), but the marker code is still relevant.

The actual call used for opening tooltips: IModelApp.viewManager.selectedView?.[openToolTip](https://www.itwinjs.org/reference/core-frontend/views/screenviewport/opentooltip/)

Someone from my team is currently working on a cleaner sample to demonstrate this. We'll post here when ready. 
Stay tuned.

**Reply by ajay-toobler** [reply]
Thank you so much @roopksaini  and @pmconne for your assistance.

**Reply by ajay-toobler** [reply]
@pmconne  , "how to use @MichaelBelousov 's article" u shared.
I have implemented `useMarker` that inside array of `uiProviders` of `Viewer`.

When I am using it. i am getting error like 
```
consuming the MarkerDecorationContext is invalid without a provider
```

When I am wrap it with `IModelJsViewProvider` , getting error like 
```
cannot read properties of undefined (reading 'addDecorator')
```

**Reply by pmconne** [reply]
If you want help debugging your code, you will need to share your code.
Creating a [sandbox](https://www.itwinjs.org/sandboxes/) is one easy way to do that.

**Reply by ajay-toobler** [reply]
Can't import this package "@bentley/imodel-react-hooks" inside sandbox . 

Can u Help me to guide how to create UI element like small Overlay PopUp near to the UI-element the user clicked.

**Reply by aruniverse** [reply]
To use the `IModelJsViewProvider` component from [@itwin/imodel-react-hooks](https://www.npmjs.com/package/@itwin/imodel-react-hooks), [IModelApp](https://www.itwinjs.org/reference/core-frontend/imodelapp/imodelapp/) needs to be initialized first. The `Viewer` component is what usually takes care of this when it gets mounted, but if you're passing the Viewer component as a child to the the IModelJsViewProvider then theres a chicken or the egg issue. To get around this, you should use the `useWebViewerInitializer` hook from `@itwin/web-viewer-react` before rendering the IModelJsViewProvider and Viewer.


```ts
  useWebViewerInitializer({
	...props
  });

  return (
    <IModelJsViewProvider>
      <Viewer
        {...props}
      />
    </IModelJsViewProvider>
  )
```

**Reply by ajay-toobler** [reply]
when i did exactly as u said the screen goes blank . @aruniverse can u show me one working demo.

```tsx

 const props = {
    iTwinId: iTwinId ?? "",
    iModelId: iModelId ?? "",
    authClient: authClient,
    viewCreatorOptions: viewCreatorOptions,
    enablePerformanceMonitors: true, // see description in the README (https://www.npmjs.com/package/@itwin/web-viewer-react)
    onIModelAppInit: onIModelAppInit,
    uiProviders: [
      new ViewerNavigationToolsProvider(),
      new ViewerContentToolsProvider({
        vertical: {
          measureGroup: false,
        },
      }),
      new ViewerStatusbarItemsProvider(),
      new TreeWidgetUiItemsProvider(),
      new PropertyGridUiItemsProvider({
        enableCopyingPropertyText: true,
      }),
      new MeasureToolsUiItemsProvider(),
      new PresentationProviderWidget(),
    ],
    onIModelConnected: onIModelConnected,
  };

useWebViewerInitializer({
    ...props,
  });

  return (
    <div className="viewer-container">
      {!accessToken && (
        <FillCentered>
          <div className="signin-content">
            <ProgressLinear indeterminate={true} labels={["Signing in..."]} />
          </div>
        </FillCentered>
      )}
      <IModelJsViewProvider>
        <Viewer {...props} />
      </IModelJsViewProvider>
    </div>
  );
```

**Reply by aruniverse** [reply]
This isnt a complete sample, but this pretty much what i have, let me know if you run into issues with it:
https://gist.github.com/aruniverse/7e0c79413fc9b9acdad204b6594add53

**Reply by ajay-toobler** [reply]
I have done exactly as u given. The error is inside `IModelJsViewProvider`.  This is the error occuring :
![image](https://user-images.githubusercontent.com/109501190/192690547-2bfdceb0-282b-4ff5-a4fa-e768751f28ba.png)

**Reply by ajay-toobler** [reply]
It would be great if someone share some sandbox about `useMarker`.

**Reply by aruniverse** [reply]
This is by no means a perfect example, but it does work. [See](https://github.com/aruniverse/mytwin-viewer-samples)

**Comment by claudiaareneee** [comment]
Hi! I created a sandbox that shows how to dynamically update a tooltip on a marker.

Check it out here: [Dynamic Marker Tooltips](https://www.itwinjs.org/sandboxes/ClaudiaNelson/Dynamic%20Marker%20Tooltips)

The important things that I found that keep the tooltip from flashing are: 

- Check that the tooltip is open before updating it (see MarkerPinDecorator.ts, lines 46 - 50)
```
if (IModelApp.notifications.isToolTipOpen && this._tooltipViewPoint)
      this.updateTooltip(this._toolTipNumber);
```

- Set the tooltip HTML using the openToolTip -- I wrap this in an if statement to make sure that either my mock data has changed or the tooltip needs to be open before calling this. (see MarkerPinDecorator.ts, lines 54 - 64)
```
const tooltip = `Mock Data: ${this._toolTipNumber}`;
const div = document.createElement("div");

div.innerHTML = tooltip;
this.title = div;
IModelApp.viewManager.selectedView!.openToolTip(this.title!, this._tooltipViewPoint, this.tooltipOptions);
```
